Simulation-based calibration

avi-stripe opened the pull request:
This implements Simulation-based calibration as described in https://arxiv.org/abs/1804.06788 and http://andrewgelman.com/2018/04/18/better-check-yo-self-wreck-yo-self/ . The API works like this:

* you construct a `SBC`instance using one of the two `apply` methods on the `SBC` companion object. The univariate form is `SBC(priorDist){x => buildNoiseDist(x)}` where `priorDist` is a `Continuous` distribution, `x` is assumed to be drawn from `priorDist`, and `buildNoiseDist` returns any `Distribution` parameterized by `x`, which will be used first to generate synthetic data and then as a likelihood function for that data.
* you start with the `prepare` method on `SBC` which will return a `SBC.Run` instance after running a single 4-chain diagnostic trace on the model; this lets you inspect the rHat for convergence before proceeding, and inspect how much SBC needs to thin your trace to get a sufficiently large effective sample size for running the calibration
* you can then call `simulate` (for programmatic use) or, better, `animate` (for interactive use) to run the actual calibration, which will run a large number of traces and check the results for uniformity per the referenced paper. Both of these methods take a `logBins` parameter, which must be between 1 and 7, and will produce histograms with 2^logBins bins. The `animate` version will animate the histograms growing in your terminal (which can take quite a long time, so this acts as a kind of progress bar), as well as giving you other diagnostic info like the rHat, and drawing confidence bounds on the histogram bars so you can see  whether the non-uniformity is within reasonable limits.

To see this in action, try the `SBCNormal` example.

The next step would be to use this from the test cases, but I plan to do that as a separate PR.

roban-stripe commented on rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala:
```diff
@@ -0,0 +1,173 @@
+      posterior.map(_._1.generator.repeat(syntheticSamples)).get
+    val model = posterior.flatMap {
+      case (d, r) =>
+        d.fit(syntheticValues).map { _ =>
```
The computation graph (I don't know what to call it) style gets a little mind-bending for me here, since at first it looks like we have the value of `r` before we actually do the `fit` that constrains it. But I figured out what's happening: the `Real` `r` doesn't have a numerical value within this case statement -- we're just building up DAG that links the value of `r` with the fit of `d` to `syntheticValues`.

avi-stripe replied:
Yep, exactly.

roban-stripe commented on rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala:
```diff
@@ -0,0 +1,173 @@
+
+    val maxRHat = diag.map(_.rHat).max
+    val minEffectiveSampleSize = diag.map(_.effectiveSampleSize).min
+    val rawRank = samples.tail.count { n =>
```
Why the `tail`?

avi-stripe replied:
We have some `L` which is the number of possible values we want `rawRank` to be able to take on. (As it happens we hard code that to `128`). Because the true value can be before the first sample, or after the last sample (or anywhere in between), this means we only want to compare it against `L-1` different samples. But we're producing our samples from 4 different chains, which means we'd need to sample `(L-1)/4` from each chain. Since we can't sample fractional samples, and since we know that `L` in (and always should be) evenly divisible by 4, we just sample `L/4` (= 32) from each chain and then throw one out to end up with `L-1`.

avi-stripe replied:
Also we throw out the first one in particular because it's the least likely one to be converged.

roban-stripe replied:
Ah, thanks. That fractional sample subtlety escaped me.

roban-stripe commented on rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala:
```diff
@@ -0,0 +1,173 @@
+      println("\u001b[1000D") //move left
+      println(s"\u001b[${bins + 3}A") //move up
+      val remaining = formatMillis(millisRemaining)
+      println(s"Repetition $rep/$reps. Estimated time remaining: $remaining")
```
This dynamic plot is really great! Could we add a line here explaining that the `[` and `]` bracket the expected 0.5-99.5th percentiles in each bin?

roban-stripe replied:
Oh, and you mentioned printing rHat, but I don't see it. Maybe print rHat and any other interesting diagnostics?

avi-stripe replied:
Will do.

avi-stripe commented on rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala:
```diff
@@ -0,0 +1,173 @@
+      (fn(l.head), l.head)
+    }
+
+  val Samples = 128
```
It's a constant. I could have used `SAMPLES`. I'm honestly not sure what's considered better scala style.

avi-stripe commented on rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala:
```diff
@@ -0,0 +1,173 @@
+
+  def prepare(sampler: Sampler, warmupIterations: Int, syntheticSamples: Int)(
+      implicit rng: RNG): Run = {
+    val initialThin = 2
```
I don't have any better justification than that :) Basically, my expectation is that most models will benefit from some thinning, so I might as well start there.

roban-stripe replied:
sg

seantalts replied:
You should be able to calculate how much to thin based on the number of effective samples. There's an example in the paper in section 4.

avi-stripe replied:
@seantalts yes, thanks - that's what I was trying to do here: https://github.com/stripe/rainier/pull/107/files#diff-c279ee7bd4b82e7ee0df025715f7fddfR22 . This is just about the thinning value to use for the very first run from which we derive the effective sample size in the first place. (Unlike in the paper, we only look at that once, and then use the same thinning factor for all subsequent runs. This was an attempt to be slightly more efficient but may come at an unacceptable correctness cost, I'd be curious to hear thoughts there).

seantalts replied:
Gotcha. That didn't work for us - you can get very different effective samples per sample rates across different prior draws.

avi-stripe replied:
Ok, makes sense. I'll fix that. Thanks!

avi-stripe commented on rainier-core/src/main/scala/com/stripe/rainier/core/SBC.scala:
```diff
@@ -0,0 +1,173 @@
+
+object SBC {
+  def apply[T](priors: Seq[Continuous])(
+      fn: Seq[Real] => (Distribution[T], Real)): SBC[T] = {
```
Exactly correct.

avi-stripe replied:
I'll put that as a comment :)

avi-stripe commented:
@roban-stripe thanks for the review. I fixed the bug that was hiding the rHat diagnostics, added a note about the [] quantile markers, and added a comment.

roban-stripe commented:
Great, LGTM so far. Happy to re-review when you make the change to the Neff calculations (if you want to do that in this PR).

roban-stripe commented:
Nice. Love the new animation with more info and red/green lines!

LGTM!
